Amplicon Sequencing Resources
=============================

Review Articles
---------------
- [Comparing bioinformatic pipelines for microbial 16S rRNA amplicon sequencing](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0227434)
- [Denoising the Denoisers: an independent evaluation of microbiome sequence error-correction approaches](https://peerj.com/articles/5364/)
- [Exact sequence variants should replace operational taxonomic units in marker-gene data analysis](https://www.nature.com/articles/ismej2017119)

Methods
-------
- [OptiClust, an Improved Method for Assigning Amplicon-Based Sequence Data to Operational Taxonomic Units](https://msphere.asm.org/content/2/2/e00073-17.short)
- [DADA2: High resolution sample inference from Illumina amplicon data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4927377/)
- [Deblur Rapidly Resolves Single-Nucleotide Community Sequence Patterns](https://msystems.asm.org/content/2/2/e00191-16.short)
- [UNOISE2: improved error-correction for Illumina 16S and ITS amplicon sequencing](https://doi.org/10.1101/074252)

Tutorials
---------
- [DADA2](https://benjjneb.github.io/dada2/tutorial.html)
 - [Sample Pooling](https://benjjneb.github.io/dada2/pool.html)
 - [Psudopooling](https://benjjneb.github.io/dada2/pseudo.html)
 - [Big Data Processing](https://benjjneb.github.io/dada2/bigdata.html)

- mothur
 - [MiSeq SOP](https://mothur.org/wiki/miseq_sop/)
 - [mothur Analysis Examples](https://mothur.org/wiki/analysis_examples/)
 - [16S Microbial Analysis with mothur (extended)](https://training.galaxyproject.org/training-material/topics/metagenomics/tutorials/mothur-miseq-sop/tutorial.html)

- QIIME2
 - [Importing Data](https://docs.qiime2.org/2021.2/tutorials/importing/)
 - [Moving Pictures](https://docs.qiime2.org/2021.2/tutorials/moving-pictures/)
 - [Atacama Soil Microbiome](https://docs.qiime2.org/2021.2/tutorials/atacama-soils/)
 - [Paired-End Analysis](https://github.com/qiime2/docs/blob/master/source/tutorials/read-joining.rst)
 